File: www/hex/make_hex.py
"""Generate point-up hex stickers for the site in the site palette. Seeded; rerun to regenerate."""
import random, re
PAPER,INK,SOFT,TEAL,INDIGO,RED='#FBFAF7','#1C1B22','#4B4956','#21918C','#3B0F5C','#B3261E'
W,H=200,231
HEX=[(100,4),(196,60),(196,171),(100,227),(4,171),(4,60)]
POLY=' '.join(f'{x},{y}' for x,y in HEX)

def manhattan(seed=7):
    random.seed(seed)
    x0,x1,floor,top=30,170,168,88          # plot box: middle band of the hex
    n_chr=12; cw=(x1-x0)/n_chr
    towers={1:0.95,4:0.62,7:1.0,9:0.5}      # chromosome -> relative tower height
    out=[]
    for c in range(n_chr):
        col=TEAL if c%2==0 else INDIGO
        cx0=x0+c*cw
        # dense floor: many small dots, heights decaying so the skyline has a shaggy top
        for _ in range(140):
            x=cx0+random.uniform(0.6,cw-0.6)
            h=min(random.expovariate(1/6.5),26)
            out.append(f'<circle cx="{x:.1f}" cy="{floor-h:.1f}" r="1.25" fill="{col}"/>')
        # a few scattered mid-height dots
        for _ in range(6):
            x=cx0+random.uniform(1,cw-1); h=random.uniform(26,36)
            out.append(f'<circle cx="{x:.1f}" cy="{floor-h:.1f}" r="1.25" fill="{col}"/>')
        if c in towers:
            tx=cx0+cw/2; th=(floor-top)*towers[c]
            for _ in range(38):
                x=tx+random.gauss(0,0.9); h=random.uniform(20,th)
                out.append(f'<circle cx="{x:.1f}" cy="{floor-h:.1f}" r="1.35" fill="{col}"/>')
            out.append(f'<circle cx="{tx:.1f}" cy="{floor-th:.1f}" r="1.6" fill="{col}"/>')
    thresh=floor-22
    axes=f'''
    <line x1="{x0-3}" y1="{floor+1}" x2="{x1+3}" y2="{floor+1}" stroke="{INK}" stroke-width="1.8" stroke-linecap="round"/>
    <line x1="{x0-3}" y1="{floor+1}" x2="{x0-3}" y2="{top-6}" stroke="{INK}" stroke-width="1.8" stroke-linecap="round"/>
    <line x1="{x0}" y1="{thresh}" x2="{x1}" y2="{thresh}" stroke="{RED}" stroke-width="1.4" stroke-dasharray="3.5 2.5"/>
    <text x="{(x0+x1)/2}" y="{floor+12}" text-anchor="middle" font-family="JetBrains Mono, Menlo, monospace" font-size="6.5" letter-spacing="0.8" fill="{SOFT}">CHROMOSOME</text>
    <text x="{x0-8}" y="{(floor+top)/2}" text-anchor="middle" transform="rotate(-90 {x0-8} {(floor+top)/2})" font-family="JetBrains Mono, Menlo, monospace" font-size="6.5" letter-spacing="0.8" fill="{SOFT}">-LOG10(P)</text>'''
    title=f'''
  <text x="100" y="60" text-anchor="middle" font-family="JetBrains Mono, Menlo, monospace" font-weight="500" font-size="9" letter-spacing="1.6" fill="{INK}">WHERE I CAME FROM</text>
  <text x="100" y="74" text-anchor="middle" font-family="Cormorant Garamond, Georgia, serif" font-style="italic" font-size="12" fill="{SOFT}">two careers, stapled</text>'''
    return f'''<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 {W} {H}" width="{W}" height="{H}" role="img" aria-label="Hex sticker: a cartoon Manhattan plot, four towers rising above a dashed significance line">
  <defs><clipPath id="hexclip"><polygon points="{POLY}"/></clipPath></defs>
  <polygon points="{POLY}" fill="{PAPER}" stroke="{INK}" stroke-width="7" stroke-linejoin="round"/>
  <g clip-path="url(#hexclip)">{axes}
    {''.join(out)}
  </g>{title}
</svg>'''

if __name__=='__main__':
    full=manhattan()
    open('www/hex/origins.svg','w').write(full)
    # icon: no title text; plot scaled up about the hex centre
    icon=re.sub(r'\n  <text x="100".*?</text>','',full,flags=re.S)
    icon=icon.replace('<g clip-path="url(#hexclip)">','<g clip-path="url(#hexclip)"><g transform="translate(100 118) scale(1.28) translate(-100 -118)">').replace('\n  </g>','\n  </g></g>',1)
    icon=re.sub(r'aria-label="[^"]*"','aria-label=""',icon)
    open('www/hex/origins-icon.svg','w').write(icon)
    print('wrote origins.svg, origins-icon.svg')

File: www/hex/make_hex.R
# make_hex.R -- point-up hex stickers for petertanksley.github.io
#
# One function per sticker returns a list of ggplot layers drawn in a 200 x 231
# "sticker coordinate" space (x right, y down, like SVG). hex_sticker() wraps the
# layers in the hex border, optional title, and writes an SVG via svglite.
#
# Usage:  Rscript www/hex/make_hex.R        (from the site root)
# Adds a sticker: write a new art_*() function and add two write_hex() calls below.

suppressPackageStartupMessages({
  library(ggplot2)
  library(dplyr)
  library(svglite)
})

# ---- palette (matches theme.scss) --------------------------------------------
PAL <- c(paper = "#FBFAF7", ink = "#1C1B22", soft = "#4B4956",
         teal = "#21918C", indigo = "#3B0F5C", red = "#B3261E", brass = "#C9A227")

W <- 200; H <- 231
HEX <- data.frame(x = c(100, 196, 196, 100, 4, 4),
                  y = c(4, 60, 171, 227, 171, 60))

# sticker coords (y down) -> ggplot coords (y up)
flip <- function(y) H - y
px   <- function(pt) pt / ggplot2::.pt          # font size in "px" -> geom_text size

# ---- art: Manhattan plot with a police cap on the y-axis ---------------------
# "Where I came from": criminology + behavior genetics, two careers stapled.
art_origins <- function(seed = 7, n_chr = 8) {
  set.seed(seed)
  x0 <- 58; x1 <- 172; floor <- 178; top <- 108         # plot box, shifted right
  cw <- (x1 - x0) / n_chr
  towers <- c(`1` = 0.95, `3` = 0.6, `5` = 1.0, `6` = 0.5) # chromosome -> rel. height

  chr <- 0:(n_chr - 1)
  dots <- bind_rows(lapply(chr, function(c) {
    col <- if (c %% 2 == 0) PAL["teal"] else PAL["indigo"]
    cx0 <- x0 + c * cw
    fl  <- data.frame(x = cx0 + runif(150, 0.6, cw - 0.6),
                      h = pmin(rexp(150, 1 / 6.5), 26), r = 1.25)
    md  <- data.frame(x = cx0 + runif(6, 1, cw - 1), h = runif(6, 26, 36), r = 1.25)
    tw  <- NULL
    if (as.character(c) %in% names(towers)) {
      tx <- cx0 + cw / 2; th <- (floor - top) * towers[[as.character(c)]]
      tw <- data.frame(x = tx + rnorm(38, 0, 0.9), h = runif(38, 20, th), r = 1.35) |>
        bind_rows(data.frame(x = tx, h = th, r = 1.6))
    }
    bind_rows(fl, md, tw) |> mutate(col = col)
  })) |> mutate(y = floor - h)

  thresh <- floor - 22
  ax <- x0 - 8                                           # y-axis x position

  # police cap perched on top of the y-axis: visor, crown, band, badge
  cap_cy <- top - 14
  crown <- data.frame(x = ax + c(-13, -11, -6, 0, 6, 11, 13, 11, -11),
                      y = cap_cy + c(6, -1, -6, -8, -6, -1, 6, 9, 9))
  band  <- data.frame(xmin = ax - 13, xmax = ax + 13, ymin = cap_cy + 6, ymax = cap_cy + 11)
  visor <- data.frame(x = ax + c(-16, 16, 13, -13), y = cap_cy + c(11, 11, 16, 16))

  list(
    # axes
    annotate("segment", x = ax, xend = x1 + 3, y = flip(floor + 1), yend = flip(floor + 1),
             colour = PAL["ink"], linewidth = 0.55, lineend = "round"),
    annotate("segment", x = ax, xend = ax, y = flip(floor + 1), yend = flip(cap_cy + 15),
             colour = PAL["ink"], linewidth = 0.55, lineend = "round"),
    # genome-wide significance line
    annotate("segment", x = x0, xend = x1, y = flip(thresh), yend = flip(thresh),
             colour = PAL["red"], linewidth = 0.42, linetype = "22"),
    # the skyline
    geom_point(data = dots, aes(x, flip(y), colour = I(col), size = I(r * 0.62)), stroke = 0),
    # cap
    geom_polygon(data = visor, aes(x, flip(y)), fill = PAL["ink"]),
    geom_polygon(data = crown, aes(x, flip(y)), fill = PAL["indigo"]),
    geom_rect(data = band, aes(xmin = xmin, xmax = xmax, ymin = flip(ymax), ymax = flip(ymin)),
              fill = PAL["ink"]),
    annotate("point", x = ax, y = flip(cap_cy + 1), colour = PAL["brass"], size = 1.9),
    # axis labels
    annotate("text", x = (x0 + x1) / 2, y = flip(floor + 12), label = "CHROMOSOME",
             family = "mono", size = px(6.5), colour = PAL["soft"]),
    annotate("text", x = ax - 8, y = flip((floor + top) / 2), label = "-LOG10(P)", angle = 90,
             family = "mono", size = px(6.5), colour = PAL["soft"])
  )
}

# ---- sticker wrapper ----------------------------------------------------------
hex_sticker <- function(art, title = NULL, subtitle = NULL) {
  p <- ggplot() +
    geom_polygon(data = HEX, aes(x, flip(y)), fill = PAL["paper"], colour = NA)
  p <- p + art +
    geom_polygon(data = HEX, aes(x, flip(y)), fill = NA, colour = PAL["ink"],
                 linewidth = 2.2, linejoin = "round")
  if (!is.null(title))
    p <- p + annotate("text", x = 100, y = flip(60), label = title, family = "mono",
                      fontface = "bold", size = px(9), colour = PAL["ink"])
  if (!is.null(subtitle))
    p <- p + annotate("text", x = 100, y = flip(74), label = subtitle, family = "serif",
                      fontface = "italic", size = px(12), colour = PAL["soft"])
  p + coord_fixed(xlim = c(0, W), ylim = c(0, H), expand = FALSE, clip = "on") +
    theme_void() +
    theme(plot.margin = margin(0, 0, 0, 0), plot.background = element_blank())
}

write_hex <- function(p, path, aria = "") {
  svglite(path, width = W / 72, height = H / 72, bg = "transparent",
          system_fonts = list(mono = "Menlo", serif = "Georgia"))
  print(p); dev.off()
  # swap in the site's web fonts (browser falls back to the system names if unavailable)
  s <- readLines(path, warn = FALSE) |> paste(collapse = "\n")
  s <- gsub('font-family: ?["\']?Menlo["\']?;',  'font-family: "JetBrains Mono", Menlo, monospace;', s)
  s <- gsub('font-family: ?["\']?Georgia["\']?;', 'font-family: "Cormorant Garamond", Georgia, serif;', s)
  s <- sub("<svg ", sprintf("<svg role='img' aria-label='%s' ", aria), s)
  writeLines(s, path)
  invisible(path)
}

# ---- build ---------------------------------------------------------------------
if (sys.nframe() == 0) {
  dir <- "www/hex"
  write_hex(hex_sticker(art_origins(), "WHERE I CAME FROM", "two careers, stapled"),
            file.path(dir, "origins.svg"),
            aria = "Hex sticker: a cartoon Manhattan plot with a police cap on the y-axis")
  write_hex(hex_sticker(art_origins()), file.path(dir, "origins-icon.svg"))
  cat("wrote origins.svg, origins-icon.svg\n")
}
